✨ add SSO support (LDAP, OIDC)

Rikpat opened the pull request:
### Category
Feature

### Overview
Changed structure of next-auth to allow the use of multiple providers. Implemented LDAP and OIDC Providers (tested on lldap + authelia stack). 
I've created this for my own use and integration with Authelia for my home lab. I've added LDAP as there is an existing issue requesting this and I already had Authelia synced with lldap. 
I've reworked how the adapter is used in next-auth, so the providers can be adapter agnostic (I have not touched password auth or db structure)

Multiple providers are supported at the same time. Having a user created by ldap prevents signing in using password as there would be no password hash, but if the user was created with password first (by invite) would be accessible by ldap. 
I suggest using just one provider at a time, as this can cause security issues. OIDC won't register to an existing email, as it uses next-auths model of having account and user separately and it has to be linked. 
For any future multi-provider setups I suggest using this approach for credential providers as well.

Another thing that should be done and I didn't do to keep the PR simple is modify the onboarding process and middleware. It should redirect to login, and if only credentials provider is enabled and no user is found the login page itself should redirect to onboarding. This would require server session from next-auth, to not redirect when user is logged in, but in current version it's not feasible with db session (and next-auth middleware is not working because of [this caveat](https://next-auth.js.org/configuration/nextjs#caveats)). 
For now, onboarding an oidc or ldap user requires manually going to /auth/login

### Issue Number
Related issue: #747

### New Vars

Most important variable is **AUTH_PROVIDER** a comma separated list of providers, supported values: credentials, ldap, oidc

Variables for LDAP
- **AUTH_LDAP_URI**: URI of your LDAP server
- **AUTH_LDAP_BIND_DN**: dn of account used for user and group querying
- **AUTH_LDAP_BIND_PASSWORD**: password of account used for user and group querying
- **AUTH_LDAP_BASE**: base of your ldap accounts
- **AUTH_LDAP_USERNAME_ATTRIBUTE**: z.string().default('uid'),
- **AUTH_LDAP_GROUP_CLASS**: z.string().default('groupOfUniqueNames'),
- **AUTH_LDAP_GROUP_MEMBER_ATTRIBUTE**: z.string().default('member'),
- **AUTH_LDAP_GROUP_MEMBER_USER_ATTRIBUTE**: z.string().default('dn'),
- **AUTH_LDAP_ADMIN_GROUP**: z.string().default('admin'),
- **AUTH_LDAP_OWNER_GROUP**: z.string().default('admin'),

Variables for OIDC (self describing)
- AUTH_OIDC_CLIENT_ID
- AUTH_OIDC_CLIENT_SECRET
- AUTH_OIDC_URI
- AUTH_OIDC_CLIENT_NAME: Custom display name for the provider
- AUTH_OIDC_ADMIN_GROUP
- AUTH_OIDC_OWNER_GROUP

### Screenshot
Updated login page with all providers
![image](https://github.com/ajnart/homarr/assets/33869814/aad8b5ac-fb82-47a8-9589-c98a8a405033)

Updated login page with oidc only
![image](https://github.com/ajnart/homarr/assets/33869814/e82440c6-3062-41dc-b9ce-c2dfc5868a70)

manuel-rw commented:
Hi, wow this is pretty cool.
Thank you for taking the time and doing this.
However, we will not merge this for 0.14.0 as we don't want to introduce any potentially unstable and not tested features.
Is that okay for you?
I think it should be doable to merge this for 0.14.1 though - so right after it.

manuel-rw commented:
I'm really hyped for this!
@Meierschlumpf @ajnart

ajnart commented:
Hey @Rikpat , thanks a lot for your contribution !

This will be an awesome feature to have in homarr. However, I am not very familiar with ldap and oidc, could you please help me set-up a replicable test environment in order to test your PR ? 
I am not sure how to setup authelia correctly in order to test it out.

As a sidenote, it's true that we have to re-think the onboarding process, it was meant to be some kind of migration page only for jumping on to the new version, because there is a new mount point needed, but the next versions should have a smaller and more refined onboarding page for new installs.

Rikpat commented:
Hi, I'm definitely looking forward to improving the authentication and roles in Homarr, as my goal was to have a homepage for my  home network with multiple dashboards and access levels (me as admin, roommates as users, guests) and Homarr looked the closest to having these features.

I'm also not that familiar with LDAP but there already was an issue about including the support, so I tried. I'm not sure if it works with all LDAP providers, but it's modular enough with the env variables. I'm running lldap as backend for my Authelia instance, so I got it working with it. 

You can try [running lldap in docker](https://github.com/lldap/lldap#with-docker). LDAP integration requires having an account that can list users and groups (for example the default admin account). 

For OIDC any provider supporting the protocol should work, I've tested it just with Authelia, but anything from Azure AD through Google to self hosted stuff like Authelia or Authentik should work. 

Authelia requires the OIDC to be enabled and having a client for Homarr (this is my config)
```yaml
identity_providers:
  oidc:
    ...
    clients:
    - id: homarr
      secret: ---
      public: false
      authorization_policy: one_factor
      redirect_uris:
        - https://example.com/api/auth/callback/oidc
        - http://localhost:3000/api/auth/callback/oidc
      scopes:
        - openid
        - groups
        - profile
        - email
      userinfo_signing_algorithm: none
      consent_mode: implicit # self hosted
```

For the simplest Authelia deployment for testing I would suggest just using file user management like this [lite or local configuration](https://www.authelia.com/integration/deployment/docker/#lite).

ajnart commented:
Unfortunately I still haven't found time to review this as I'm dreading documenting myself regarding authelia, and LDAP and setting them up locally for testing purposes. I have no prior experiences with these appart from my local authelia setup using only the default templates 
Could you please write related documentation articles about this feature so that it can be put into the docs ? (And so that I can follow these docs to test if it works on my side)

You can fork https://github.com/ajnart/homarr-docs/pull/83 (`auth-update` branch) to get started with documentation, then you could add your article to a `login.mdx` page in /managment for example. We still have to write some documentation about the login process. 
You'd also need to document the environment variables in the correct page for them on the docs.

Would you mind doing that ? We really need to release documentation 1 to 1 with the source code, especially for features like these otherwise new users wanting to use these features will be lost in trying to set them up.

manuel-rw commented:
We are going to properly review this in the following days. We want to get the 0.14 release right and as stable as possible. Thanks for your understanding!

ajnart commented:
Hello @Rikpat , I’m sorry but I have absolutely no clue how to setup ldap and oidc with authelia locally to test your PR. I was able to start authelia with the local template but it didn’t work inside of homarr for me. 
Im afraid this PR will stay stale until I can find a way to set it up for testing purposes on my local pc.

are you able to assist me via discord by any chance ? Please send me a DM there if so

Polymeta commented:
Big fan of the PR. Can also help with testing where required!

Also for testing the OIDC component of this PR, you can also hook it up to a remote identity server (google, Microsoft, okta, etc.) instead of trying to host your own authelia if you are struggling with that.

manuel-rw commented:
I will be reviewing this over the weekend.
Expect tons of comments from me.
Would love to see this part of ``0.14.3`` or ``0.14.4``.
We originally planned to do this ourselves but this contribution is awesome.
I'll keep you up to date 😉

Rikpat commented:
Hello, sorry about late reply, but I'm in a middle of apartment renovation and when I'm not working I'm usually there. If you still need any help, I should be available on Monday.

manuel-rw commented:
Wasn't able to set up the lite configuration - it failed for some reason.
But I'll try again this weekend - I should have more time for it now. Sorry 😅

manuel-rw commented:
Thank you for the offer. I'll try to run myself for now. I want to set up Authelia anyway as soon as Homarr supports it (I have too many different login screens in my home lab - SSO is the way!).
I'll let you guys know in case I have any questions or problems.
I did not have a look at all the code yet - I'll post a bigger review within the next days.

manuel-rw commented on src/utils/auth/index.ts:
```diff
@@ -0,0 +1,56 @@
+
+// const availableProviders = {
+//     credentials: "./credentials",
+//     ldap: "./ldap"
```
Remove commented out code

manuel-rw reviewed (changes requested):
Code looks good to me. I just have troubles getting Authelia to run properly.
- I had to configure ``identity_providers.oidc.issuer_private_key`` to a RSA key. This seems to work.
- I can reach authelia now using a local DNS record on my Pihole on https://authelia.example.com.
- I get a certificate error when I try to authenticate:
![image](https://github.com/ajnart/homarr/assets/30572287/6f9e2372-906b-4309-be83-7d79a2f12f14)

I had to go through quite a few manual steps to get it running (including not-documented steps in the Authelia docs).
Do you have an idea why I am getting the certificate error?

manuel-rw commented on src/env.js:
```diff
@@ -28,6 +30,37 @@ const env = createEnv({
+          AUTH_OIDC_CLIENT_SECRET: z.string(),
+          AUTH_OIDC_URI: z.string().url(),
+          // Custom Display name, defaults to OIDC
+          AUTH_OIDC_CLIENT_NAME: z.string().default('OIDC'),
```
I was able to set ``AUTH_OIDC_CLIENT_ID`` even with unset ``AUTH_PROVIDER``. Is this intentional?

manuel-rw commented on src/pages/auth/login.tsx:
```diff
@@ -108,36 +114,71 @@ export default function LoginPage({
+                      name="ldap"
+                      onClick={() => form.setFieldValue('provider', 'ldap')}
+                    >
+                      {t('form.buttons.submit')} - LDAP
```
Should be translated -> Not all languages are LTR

manuel-rw commented on src/server/db/schema.ts:
```diff
@@ -2,7 +2,9 @@ import { InferSelectModel, relations } from 'drizzle-orm';
 
-export const users = sqliteTable('user', {
+// workaround for typescript check in adapter
+// preferably add email into credential login and make email non-nullable here
```
Can you elaborate why this workaround is needed?

Tagaishi replied:
If this is a necessity, force the email in the existing credential system. IMO.
At first it wasn't even supposed to be in, to make the login update as simple as possible, but ended up adding it because we thought it would be useful in the future.
If it's a necessity, then it's fine to make it so, we just didn't have a reason until now since it's all new.

manuel-rw commented on src/utils/auth/index.ts:
```diff
@@ -0,0 +1,56 @@
+
+export const providers: (CredentialsConfig | OAuthConfig<any>)[] = []
+
+if (env.AUTH_PROVIDER.includes('ldap')) providers.push((await import("./ldap")).default)
+if (env.AUTH_PROVIDER.includes('credentials')) providers.push((await import("./credentials")).default)
+if (env.AUTH_PROVIDER.includes('oidc')) providers.push((await import("./oidc")).default)
```
I assume this is what you needed the webpack configuration for?

Rikpat replied:
yes. It was to not run the initialization of all providers if not necessary

manuel-rw commented on src/utils/auth/ldap.ts:
```diff
@@ -0,0 +1,146 @@
+
+
+// Helper types for infering properties of returned search type
+type AttributeContraint = string | readonly string[] | undefined;
```
Typo

manuel-rw commented on src/utils/auth/ldap.ts:
```diff
@@ -0,0 +1,146 @@
+        attributes: ['uid', 'mail'] as const
+      }))[0];
+
+      await ldapLogin(ldapUser.dn, data.password).then(client => client.destroy())
```
Missing ``;``?

manuel-rw commented on src/utils/auth/ldap.ts:
```diff
@@ -0,0 +1,146 @@
+      
+      Consola.log(`user ${data.name} successfully authorized`);
+
+      let user = await adapter.getUserByEmail!(ldapUser.mail);
```
Why the suppression? Can you explain?

manuel-rw commented on src/utils/auth/ldap.ts:
```diff
@@ -0,0 +1,146 @@
+        isAdmin: isAdmin,
+        isOwner: isOwner,
+      };
+    } catch (e) {
```
Use full name, no ``e``. ``i`` and ``x`` are poor common variable names.

manuel-rw commented on src/utils/auth/oidc.ts:
```diff
@@ -0,0 +1,51 @@
+  type: "oauth",
+  clientId: env.AUTH_OIDC_CLIENT_ID,
+  clientSecret: env.AUTH_OIDC_CLIENT_SECRET,
+  wellKnown: `${env.AUTH_OIDC_URI}/.well-known/openid-configuration`,
```
Is this a common standard?

Rikpat replied:
yes, the suffix is part of specification, but the env variable could technically point to directly to the configuration as it's used nowhere else

Tagaishi commented:
running credentials or ldap or both will give me this error about oidc even thought it's not part of my config.
![image](https://github.com/ajnart/homarr/assets/26098587/0fc23436-a8e1-49db-89bb-d4c9c5e5b848)

Rikpat commented:
> Code looks good to me. I just have troubles getting Authelia to run properly.
> 
> * I had to configure `identity_providers.oidc.issuer_private_key` to a RSA key. This seems to work.
> * I can reach authelia now using a local DNS record on my Pihole on https://authelia.example.com.
> * The "lite" example is pretty poorly documented in my opinion. Where is the actual database for users? (As far as I'm concerned, Authelia does not store or manage any users itself). Also 2FA was active by default? Why? 😅
> * I get a certificate error when I try to authenticate:
>   ![image](https://private-user-images.githubusercontent.com/30572287/287480055-6f9e2372-906b-4309-be83-7d79a2f12f14.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE1NTI2ODksIm5iZiI6MTcwMTU1MjM4OSwicGF0aCI6Ii8zMDU3MjI4Ny8yODc0ODAwNTUtNmY5ZTIzNzItOTA2Yi00MzA5LWJlODMtN2Q3OWEyZjEyZjE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAyVDIxMjYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4ZTVjZDU3NWQzMzVlYzYyZjk4MTk4MDdjNDY3NTRjYzBiZGUzNjg3ZTc0YjZhOGQwMjdhMTk3ZDMwODk3NjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.f0v9FWfSwqikYbdjSfvbtd5H-XVBjIXDHkDN6pQn9kI)
> 
> I had to go through quite a few manual steps to get it running (including not-documented steps in the Authelia docs). Do you have an idea why I am getting the certificate error?
> 
> @Tagaishi is now trying to run this with LDAP on his existing setup.

I think the self-signed certificate issue is related to tls certificate of the domain (https://authelia.example.com), I have my own domain for my homelab and a regular signed certificate by letsencrypt. 

Try adding your certificate to your trusted certificates, or try setting up a domain, it simplifies a lot of things (I'm using [.pp.ua](https://nic.ua/en/domains/.pp.ua) which is free and using cloudflare as free DNS, and my only record is one for DNS-01 certificate as I'm not exposing anything anywhere outside my home network).

I'll look into the rest of the comments now

Rikpat commented on src/middleware.ts:
```diff
@@ -32,12 +33,15 @@ export async function middleware(req: NextRequest) {
+  if (cachedUserCount > 0 || !(await shouldRedirectToOnboard())) {
+    // redirect to login if not logged in
+    // not working, should work in next-auth 5
+    // @see https://github.com/nextauthjs/next-auth/pull/7443
 
-  // Do not redirect if there are users in the database
-  if (!(await shouldRedirectToOnboard())) {
+    // const session = await getServerSession(); 
+    // if (!session?.user) {
+    //   return NextResponse.redirect(getUrl(req) + '/auth/login')
+    // }
```
this was moved to the other if check, I wanted to add the login redirect as when I was developing I was logging in and out a lot and not getting redirected to login page was an incovenience. Unfortunately getServerSession doesn't work in middleware for database adapters, hence the comment of next-auth 5

Rikpat commented:
> running credentials or ldap or both will give me this error about oidc even thought it's not part of my config. ![image](https://private-user-images.githubusercontent.com/26098587/287484006-0fc23436-a8e1-49db-89bb-d4c9c5e5b848.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE1NTk3NDcsIm5iZiI6MTcwMTU1OTQ0NywicGF0aCI6Ii8yNjA5ODU4Ny8yODc0ODQwMDYtMGZjMjM0MzYtYThlMS00OWRiLTg5YmItZDRjOWM1ZTViODQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAyVDIzMjQwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExODAyNGNkZDg4ZGVjZWVkNTQ4MDhjNTFiOTkzMjc3ZDU0YTFlZWI5N2I0NDdlNWViMTY0OGQwNTQxYzUzNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uaDppQ5E2esDuKKxN-2tLDvd032sdCVzONGfAwhEAUI)

Fixed the issue, sorry, I added that part as I was adding oidc and I didn't fully test it with just ldap enabled

Tagaishi commented:
Ldap works nicely. (lldap admins group is lldap_admin though, only thing to change in the vars that have a default value)

I know you already talk about not using multiple auth providers, but just having 2 users sharing emails between normal homarr database and OIDC makes it not behave nicely. Not much you can do about it though I guess but it wasn't working for the longest time because I had to make a fresh database first.
I think I read on another gh issue that you should modify if the user exists instead of creating one.

I would maybe refer to "credentials" as "native userbase" (just a bit confusing since ldap also uses credentials on the page).

Rikpat commented:
> Ldap works nicely. (lldap admins group is lldap_admin though, only thing to change in the vars that have a default value)
> 
> I know you already talk about not using multiple auth providers, but just having 2 users sharing emails between normal homarr database and OIDC makes it not behave nicely. Not much you can do about it though I guess but it wasn't working for the longest time because I had to make a fresh database first. I think I read on another gh issue that you should modify if the user exists instead of creating one.
> 
> I would maybe refer to "credentials" as "native userbase" (just a bit confusing since ldap also uses credentials on the page).

As for lldap_admin, that group allows you to edit lldap and add groups to users. Most of the time you want your users to be admins in apps but not give them access to modify other peoples access, so the best approach is to create a new group for that.

And regarding the multiple ways to sign in to a single user, I said it's possible but not recommended. Also next-auth separates 2 entities, [accounts](https://next-auth.js.org/v3/adapters/models#account) and [users](https://next-auth.js.org/v3/adapters/models#user), each user can have multiple accounts to sign in. I didn't want to modify the existing credentials provider, and I used the ldap in much the same way, but maybe, if you want to have the option of multiple ways of signin, maybe both credentials and ldap can be reimplemented to create accounts linked to a user.

I would also modify the credential provider to require and verify email somehow if using this approach as you could technically just get into somebody elses dashboards if you create a credentials account with same email as someone with other provider (this is actually covered right now, as since it's in user and that gets created with null password, you can't sign in into that account with any password).

Main benefit of multiple providers is in case of usage in some bigger environment with many users, you might want to have multiple oauth providers (eg. google + github or facebook). You can use one of these with OIDC without an issue, but for multiple, separate providers would have to be implemented. That's why I made the providers kind of pluggable and using adapter instead of database directly, so that is pluggable as well. It should be pretty straightforward to add other providers.

But this is all hypothetical as I imagine that most users run this on their homelab and are okay with a single provider and even in enterprise setting you would probably have something like AD which could just use LDAP or OIDC for all users.

manuel-rw commented on src/pages/auth/login.tsx:
```diff
@@ -108,36 +114,71 @@ export default function LoginPage({
+            )}
+            {providers.includes('oidc') && (
+              <Button mt="xs" variant="light" fullWidth onClick={() => signIn('oidc')}>
+                {t('form.buttons.submit')} - {oidcProviderName}
```
The format of the string (see my comment about LTR and RTL).

Rikpat commented:
I've done a quick workaround for auto login using Authelia (or other OIDC provider), all you have to do is supply AUTH_OIDC_AUTO_LOGIN="true". The text will need to be translated as well, but this is for showcase

Tagaishi commented:
> I've done a quick workaround for auto login using Authelia (or other OIDC provider), all you have to do is supply AUTH_OIDC_AUTO_LOGIN="true". The text will need to be translated as well, but this is for showcase

That's really nice! works well too.
You mentioned some security concerns though? If there still are, make sure to detail it in the docs clearly.

Tagaishi commented:
Caught a little something using the auto-login.

Go through the Authelia login using account 1.
Accept the prompt.
Wait for the Authelia session cookie to expire.
Reload and login using account 2.
Homarr will still be logged in account 1.

Rikpat commented:
> You mentioned some security concerns though? If there still are, make sure to detail it in the docs clearly.

That was for the REMOTE_USER header, and even then there would have to be something between your proxy and app (which would normally run on the same machine) so the risk is almost nonexistent

> Go through the Authelia login using account 1.
> Accept the prompt.
> Wait for the Authelia session cookie to expire.
> Reload and login using account 2.
> Homarr will still be logged in account 1.

yes, but you would have the same problem without auto login, the session is persisted in database, it's not logging in through authelia everytime you open homarr. If you want to change user, you first have to sign out of homarr.

lordratner commented:
I just want to thank everyone for this. LDAP integration is a big deal for those of us who use it to keep users manageable across many services. 

Cheers

jdirkse commented:
I just tested this PR against Keycloak (OIDC) and can confirm that it works for me as well

manuel-rw commented:
Hi, I want to give you an update why this hasn't been merged yet.
Our core team has been working on a major refactor of the backend.
Our current code base is polluted and not organized at all.
However, it is unclear how long this refactor will take.
We were also not sure, whether we'll use the same authentication library.
But our experiments have shown that there are no viable alternatives and that we'll stick with our current library.

@ajnart @Meierschlumpf  @Tagaishi I'd propose that we merge this anyway and copy later in the refactor. What do you think?

Tagaishi commented:
@Rikpat I tried to do the merge conflicts so we could merge it but there was a build failure.
I fixed it [here](https://github.com/ajnart/homarr/pull/1497/commits/16a4401010ae10c7bb676b70df7004b1cfc6fe67) but I'm not sure if that's how it's supposed to be as now there's also a vitest error.
Could you look into it real quick? We are merging once it's all sorted out.

manuel-rw commented:
To expand on Tag's comment: The unit tests are currently failing. As far as I see, you simply need to assert for the new properties. If you're comfortable with testing, you can create a follow up pull request adding more tests to it, as we plan to make use of tests much more in the future development of Homarr to reduce regressions.

manuel-rw commented:
Thank you again for the awesome contribution and work. This will be part of the next Homarr version 🚀💪

ajnart commented:
dont forget to copy over the changes of the PR form the old docs to the new docs if you have time :)
